File: npm/src/commands/plugins.js
'use strict'
const { Command } = require('commander')
const os = require('os')
const path = require('path')
const fs = require('fs')
const https = require('https')
const { t } = require('../i18n')

const REGISTRY_URL = 'https://raw.githubusercontent.com/kgsaran/trackfw-plugins/main/registry.yaml'
const MAX_PLUGIN_SIZE = 50 * 1024 * 1024
const REQUEST_TIMEOUT_MS = 30000

function fetchRegistry() {
  return new Promise((resolve, reject) => {
    https.get(REGISTRY_URL, (res) => {
      let data = ''
      res.on('data', chunk => { data += chunk })
      res.on('end', () => resolve(data))
    }).on('error', reject)
  })
}

function parseRegistryYAML(text) {
  const entries = []
  let current = null
  const lines = text.split('\n')
  for (const line of lines) {
    const trimmed = line.trim()
    if (!trimmed || trimmed === 'plugins:') continue
    if (trimmed.startsWith('- name:')) {
      if (current) entries.push(current)
      current = { name: trimmed.slice('- name:'.length).trim(), repo: '', description: '', tags: [] }
      continue
    }
    if (!current) continue
    if (trimmed.startsWith('repo:')) {
      current.repo = trimmed.slice('repo:'.length).trim()
    } else if (trimmed.startsWith('description:')) {
      let desc = trimmed.slice('description:'.length).trim()
      desc = desc.replace(/^"|"$/g, '')
      current.description = desc
    } else if (trimmed.startsWith('tags:')) {
      const raw = trimmed.slice('tags:'.length).trim().replace(/^\[|\]$/g, '')
      current.tags = raw.split(',').map(s => s.trim()).filter(Boolean)
    }
  }
  if (current) entries.push(current)
  return entries
}

function matchesKeyword(entry, kw) {
  const lkw = kw.toLowerCase()
  if (entry.name.toLowerCase().includes(lkw)) return true
  if (entry.description.toLowerCase().includes(lkw)) return true
  for (const tag of entry.tags) {
    if (tag.toLowerCase().includes(lkw)) return true
  }
  return false
}

function pluginsDir() {
  return path.join(os.homedir(), '.trackfw', 'plugins')
}

function platformOS() {
  if (process.platform === 'win32') return 'windows'
  if (process.platform === 'darwin') return 'darwin'
  return 'linux'
}

function platformArch() {
  if (process.arch === 'x64') return 'amd64'
  return process.arch
}

function listPlugins() {
  const dir = pluginsDir()
  fs.mkdirSync(dir, { recursive: true })
  return fs.readdirSync(dir).filter(f => fs.statSync(path.join(dir, f)).isFile())
}

async function installPlugin(repo) {
  let base = repo
  let tag = 'latest'
  const atIdx = repo.indexOf('@')
  if (atIdx !== -1) {
    base = repo.slice(0, atIdx)
    tag = repo.slice(atIdx + 1)
  }
  const pluginName = path.basename(base)
  const assetName = `trackfw-plugin-${pluginName}-${platformOS()}-${platformArch()}`
  const url = tag === 'latest'
    ? `https://github.com/${base}/releases/latest/download/${assetName}`
    : `https://github.com/${base}/releases/download/${tag}/${assetName}`

  const res = await fetch(url, { signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS) })
  if (!res.ok) throw new Error(t('errors.downloadFailed', { status: res.status, url }))
  const declaredSize = Number(res.headers.get('content-length') || 0)
  if (declaredSize > MAX_PLUGIN_SIZE) throw new Error(`Plugin exceeds ${MAX_PLUGIN_SIZE} bytes`)

  const dir = pluginsDir()
  fs.mkdirSync(dir, { recursive: true })
  const content = Buffer.from(await res.arrayBuffer())
  if (content.length > MAX_PLUGIN_SIZE) throw new Error(`Plugin exceeds ${MAX_PLUGIN_SIZE} bytes`)
  const destination = path.join(dir, pluginName)
  const temporary = path.join(dir, `.${pluginName}-${process.pid}.tmp`)
  try {
    fs.writeFileSync(temporary, content, { mode: 0o755 })
    fs.renameSync(temporary, destination)
  } finally {
    if (fs.existsSync(temporary)) fs.unlinkSync(temporary)
  }
}

function removePlugin(name) {
  if (!name || path.basename(name) !== name) throw new Error(`Invalid plugin name: ${name}`)
  const filePath = path.join(pluginsDir(), name)
  if (!fs.existsSync(filePath)) throw new Error(t('errors.pluginNotFound', { name }))
  fs.unlinkSync(filePath)
}

const cmd = new Command('plugins')
cmd.description(t('plugins.description'))

cmd.command('list')
  .description(t('plugins.list.description'))
  .action(() => {
    const plugins = listPlugins()
    if (plugins.length === 0) {
      console.log(t('plugins.list.empty'))
      return
    }
    plugins.forEach(p => console.log(p))
  })

cmd.command('add <repo>')
  .description(t('plugins.add.description'))
  .action(async (repo) => {
    try {
      console.log(t('plugins.add.installing', { repo }))
      await installPlugin(repo)
      const name = repo.split('@')[0].split('/').pop()
      console.log(t('plugins.add.success', { name }))
    } catch (err) {
      console.error(`Error: ${err.message}`)
      process.exit(1)
    }
  })

cmd.command('remove <name>')
  .description(t('plugins.remove.description'))
  .action((name) => {
    try {
      removePlugin(name)
      console.log(t('plugins.remove.success', { name }))
    } catch (err) {
      console.error(`Error: ${err.message}`)
      process.exit(1)
    }
  })

cmd.command('search <keyword>')
  .description('Search the plugin registry')
  .action(async (keyword) => {
    let entries
    try {
      const body = await fetchRegistry()
      entries = parseRegistryYAML(body).filter(e => matchesKeyword(e, keyword))
    } catch (err) {
      console.log(`Registry unavailable: ${err.message}`)
      return
    }
    if (entries.length === 0) {
      console.log(`No plugins found for "${keyword}"`)
      return
    }
    console.log(String('NAME').padEnd(30) + String('REPO').padEnd(30) + 'DESCRIPTION')
    console.log('-'.repeat(90))
    for (const e of entries) {
      console.log(e.name.padEnd(30) + e.repo.padEnd(30) + e.description)
    }
  })

module.exports = cmd

File: internal/plugins/plugins.go
package plugins

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

const RegistryURL = "https://raw.githubusercontent.com/kgsaran/trackfw-plugins/main/registry.yaml"
const maxPluginSize = 50 << 20
const maxRegistrySize = 1 << 20

var httpClient = &http.Client{Timeout: 30 * time.Second}

type RegistryEntry struct {
	Name        string
	Repo        string
	Description string
	Tags        []string
}

// parseRegistryYAML parseia o YAML flat do registry linha a linha.
// Formato esperado:
//
//	plugins:
//	  - name: trackfw-go-advanced
//	    repo: kgsaran/trackfw-go-advanced
//	    description: "Advanced Go generators"
//	    tags: [go, generators]
func parseRegistryYAML(body string) []RegistryEntry {
	var entries []RegistryEntry
	var current *RegistryEntry

	lines := strings.Split(body, "\n")
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || trimmed == "plugins:" {
			continue
		}
		if strings.HasPrefix(trimmed, "- name:") {
			if current != nil {
				entries = append(entries, *current)
			}
			current = &RegistryEntry{}
			current.Name = strings.TrimSpace(strings.TrimPrefix(trimmed, "- name:"))
			continue
		}
		if current == nil {
			continue
		}
		if strings.HasPrefix(trimmed, "repo:") {
			current.Repo = strings.TrimSpace(strings.TrimPrefix(trimmed, "repo:"))
		} else if strings.HasPrefix(trimmed, "description:") {
			desc := strings.TrimSpace(strings.TrimPrefix(trimmed, "description:"))
			// Remove aspas opcionais
			desc = strings.Trim(desc, `"`)
			current.Description = desc
		} else if strings.HasPrefix(trimmed, "tags:") {
			raw := strings.TrimSpace(strings.TrimPrefix(trimmed, "tags:"))
			// Formato: [go, generators]
			raw = strings.Trim(raw, "[]")
			parts := strings.Split(raw, ",")
			for _, p := range parts {
				tag := strings.TrimSpace(p)
				if tag != "" {
					current.Tags = append(current.Tags, tag)
				}
			}
		}
	}
	if current != nil {
		entries = append(entries, *current)
	}
	return entries
}

// matchesKeyword verifica se a entrada corresponde à keyword (name, description ou tags).
func matchesKeyword(e RegistryEntry, keyword string) bool {
	if strings.Contains(strings.ToLower(e.Name), keyword) {
		return true
	}
	if strings.Contains(strings.ToLower(e.Description), keyword) {
		return true
	}
	for _, tag := range e.Tags {
		if strings.Contains(strings.ToLower(tag), keyword) {
			return true
		}
	}
	return false
}

// Search busca no registry central por keyword (name, description, tags).
// Retorna nil, nil se registry indisponível (rede offline).
func Search(keyword string) ([]RegistryEntry, error) {
	resp, err := httpClient.Get(RegistryURL) //nolint:gosec
	if err != nil {
		return nil, fmt.Errorf("registry unavailable: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("registry returned status %d", resp.StatusCode)
	}
	bodyBytes, err := io.ReadAll(io.LimitReader(resp.Body, maxRegistrySize+1))
	if err != nil {
		return nil, fmt.Errorf("failed to read registry: %w", err)
	}
	if len(bodyBytes) > maxRegistrySize {
		return nil, fmt.Errorf("registry exceeds %d bytes", maxRegistrySize)
	}
	entries := parseRegistryYAML(string(bodyBytes))

	kw := strings.ToLower(keyword)
	var results []RegistryEntry
	for _, e := range entries {
		if matchesKeyword(e, kw) {
			results = append(results, e)
		}
	}
	return results, nil
}

// ResolveRepo: se repo parece um nome do registry (sem "/"), busca no registry e retorna repo real.
// Se já tem "/" → retorna como está (sem chamada de rede).
func ResolveRepo(nameOrRepo string) (string, error) {
	if strings.Contains(nameOrRepo, "/") {
		return nameOrRepo, nil
	}
	entries, err := Search(nameOrRepo)
	if err != nil {
		return "", fmt.Errorf("could not resolve %q from registry: %w", nameOrRepo, err)
	}
	for _, e := range entries {
		if strings.EqualFold(e.Name, nameOrRepo) {
			return e.Repo, nil
		}
	}
	return "", fmt.Errorf("plugin %q not found in registry", nameOrRepo)
}

func Dir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".trackfw", "plugins"), nil
}

func List() ([]string, error) {
	dir, err := Dir()
	if err != nil {
		return nil, err
	}
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, err
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	var names []string
	for _, e := range entries {
		if !e.IsDir() {
			names = append(names, e.Name())
		}
	}
	return names, nil
}

func Install(repo string) error {
	// Resolver nome do registry se não contém "/"
	resolved, err := ResolveRepo(repo)
	if err != nil {
		return err
	}
	repo = resolved

	// repo no formato "user/name" ou "user/name@tag"
	base := repo
	tag := "latest"
	for i, c := range repo {
		if c == '@' {
			base = repo[:i]
			tag = repo[i+1:]
			break
		}
	}
	pluginName := filepath.Base(base)
	assetName := fmt.Sprintf("trackfw-plugin-%s-%s-%s", pluginName, runtime.GOOS, runtime.GOARCH)
	var url string
	if tag == "latest" {
		url = fmt.Sprintf("https://github.com/%s/releases/latest/download/%s", base, assetName)
	} else {
		url = fmt.Sprintf("https://github.com/%s/releases/download/%s/%s", base, tag, assetName)
	}

	dir, err := Dir()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}

	resp, err := httpClient.Get(url) //nolint:gosec
	if err != nil {
		return fmt.Errorf("download failed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("download failed: HTTP %d for %s", resp.StatusCode, url)
	}
	if resp.ContentLength > maxPluginSize {
		return fmt.Errorf("download exceeds %d bytes", maxPluginSize)
	}

	dest := filepath.Join(dir, pluginName)
	f, err := os.CreateTemp(dir, "."+pluginName+"-*.tmp")
	if err != nil {
		return err
	}
	tmpPath := f.Name()
	defer os.Remove(tmpPath)

	written, err := io.Copy(f, io.LimitReader(resp.Body, maxPluginSize+1))
	if err != nil {
		f.Close()
		return err
	}
	if written > maxPluginSize {
		f.Close()
		return fmt.Errorf("download exceeds %d bytes", maxPluginSize)
	}
	if err := f.Sync(); err != nil {
		f.Close()
		return err
	}
	if err := f.Close(); err != nil {
		return err
	}
	if err := os.Chmod(tmpPath, 0755); err != nil {
		return err
	}
	return os.Rename(tmpPath, dest)
}

func Remove(name string) error {
	if name == "" || filepath.Base(name) != name {
		return fmt.Errorf("invalid plugin name %q", name)
	}
	dir, err := Dir()
	if err != nil {
		return err
	}
	path := filepath.Join(dir, name)
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return fmt.Errorf("plugin %q not found", name)
	}
	return os.Remove(path)
}
